copy_shard.c

```diff
@@ -0,0 +1,139 @@
+/*-------------------------------------------------------------------------
+ *
+ * copy_shard.c
+ *
+ * This file contains functions to copy a shard's data from a healthy
+ * placement.
+ *
+ * Copyright (c) 2014, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "c.h"
+#include "fmgr.h"
+#include "miscadmin.h"
+#include "postgres_ext.h"
+
+#include "copy_shard.h"
+#include "distribution_metadata.h"
+#include "pg_shard.h"
+
+#include <string.h>
+
+#include "access/heapam.h"
+#include "access/htup.h"
+#include "access/tupdesc.h"
+#include "executor/tuptable.h"
+#include "lib/stringinfo.h"
+#include "nodes/pg_list.h"
+#include "storage/lock.h"
+#include "utils/builtins.h"
+#include "utils/elog.h"
+#include "utils/palloc.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
+#include "utils/tuplestore.h"
+
+
+/* local function forward declarations */
+static bool CopyDataFromTupleStoreToRelation(Tuplestorestate *tupleStore,
+											 Relation relation);
+
+
+/* declarations for dynamic loading */
+PG_FUNCTION_INFO_V1(worker_copy_shard_placement);
+
+
+/*
+ * worker_copy_shard_placement implements a internal UDF to copy a table's data from
+ * a healthy placement into a receiving table on an unhealthy placement. This
+ * function returns a boolean reflecting success or failure.
+ */
+Datum
+worker_copy_shard_placement(PG_FUNCTION_ARGS)
+{
+	text *shardRelationNameText = PG_GETARG_TEXT_P(0);
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	int32 nodePort = PG_GETARG_INT32(2);
+	Oid shardRelationId = ResolveRelationId(shardRelationNameText);
+	char *shardRelationName = text_to_cstring(shardRelationNameText);
+	char *nodeName = text_to_cstring(nodeNameText);
+
```

Comment:
Two minor notes.

1. We have an extra newline between these blocks. Could we remove it?
2. To keep the distance low between variable declarations and definitions, what do you think of declaring ``placement``, ``task``, and ``selectAllQuery`` as NULL, and then defining them just before their use below?